sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py

```diff
@@ -123,7 +125,7 @@ class OperationTimeoutError(EventHubError):
     """Operation timed out."""
 
 
-class OwnershipLostError(Exception):
+class OwnershipLostError(AzureError):
```

Comment:
@YijunXieMS Was there a historical reason this wasn't EventHubError?  I'm inclined to normalize but figured I should ask.

cc @yunhaoling (no action needed, just echoing the above as a scream-test)

Reply:
@KieranBrantnerMagee , yea, technically, `OwnershipLostError` has nothing to do with the eventhub itself, it's related to load balance algorithm (claiming ownership) and blob operations. IIRC it's not used in the eventhub sdk, but used in the extensions sdks (checkpointstoreblob).

that's the reason I can remember why we didn't make it inherit from `EventHubError`.
but, I can also foresee the counter point on this "load balance, blob operation, they're all part of the event hub sdk, so it should be an EventHubError". With that being said, I now lean a bit towards making it inherit from `EventHubError`.

what's your thoughts on this?

Reply:
@yunhaoling @KieranBrantnerMagee just an FYI that the pylint error that ci is failing on below is related to this (I think):
[*pylint failure*
](https://dev.azure.com/azure-sdk/public/_build/results?buildId=578550&view=logs&j=b70e5e73-bbb6-5567-0939-8415943fadb9&t=23d29da4-00cb-5783-b17c-96bda25d9c52&l=77)
```
azure/eventhub/extensions/checkpointstoreblobaio/_blobstoragecsaio.py:189: [E1120(no-value-for-parameter), BlobCheckpointStore._claim_one_partition] No value for argument 'message' in constructor call
```
[line 189 in _blobstoragecasio.py](https://github.com/Azure/azure-sdk-for-python/blob/48ff8195424dbecbc738269ab85b9aa0072c56ab/sdk/eventhub/azure-eventhub-checkpointstoreblob-aio/azure/eventhub/extensions/checkpointstoreblobaio/_blobstoragecsaio.py#L189)
```raise OwnershipLostError()```

Not exactly sure what to do here, but if this is in fact related to the above conversation I'd be happy to modify my commits :).

Reply:
I'm sorry I completely miss the thread here.

it's because if `OwnershipLostError` inherits from `AzureError`, then `message` becomes a required parameter to the constructor.

I have updated your branch accordingly.

Reply:
I think `OwnershipLostError` could inherit from `EventHubError`, the reason why we didn't do that before might be due to it's an error triggered by the storage-blob sdk, not by eventhub when failing to claim a partition (update a blob).

But conceptually this error is still related to the eventhub workflow and it would bring consistent error-handling experience if we could make it derived from `EventHubError`.

thoughts? @annatisch @rakshith91

Reply:
A couple of considerations here:
- What information is in this error? Is there any HTTP errors? For example - if we are wrapping a StorageErrorException from the Blobs SDK, this is raised on an HTTP error, and therefore we should also consider this an HTTP error. Conclusion: it should be an AzureError derivative.
- A customer might be broken if they have written code like the following. Is that a possible scenario?
```python
try:
    # run
except (EventHubError, AzureError):
    # do something
except OwnershipLostError:
    # do something else
```
- Our guidelines state that any custom errors should be based on those in `azure-core`:
https://azure.github.io/azure-sdk/python_implementation.html#python-errors-new-exceptions

Reply:
Additional note regarding the code snippet above:
- Is there any reason a customer would handle this error _differently_ to any other EventHubError? i.e. if customers have written a single `except EventHubError` clause - is there any reason they would not want this clause to be processing OwnershipLostErrors?

Reply:
@annatisch 
- currently the `OwnershipLostError` is wrapping `ResourceModifiedError` and  `ResourceExistsError` from the storage blob, see code [here](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/eventhub/azure-eventhub-checkpointstoreblob/azure/eventhub/extensions/checkpointstoreblob/_blobstoragecs.py#L163-L173), so yes it should inherit from `AzureError`.
- you're right, in this way it breaks customers code, but the good news is that we don't have `EventHubError` inheriting from `AzureError` before, so chances are small that users would catch `EventHubError` and `AzureError` simultaneously.
- This is a fairly good point which pushes me to think that users usually can't do anything except logging when they meet with the `OwnershipLostError`, meanwhile for other eventhub errors, users might be able to take some action.


To summarize, the thinking of how users deal with the error `OwnershipLostError` vs `EventHubError` makes me now lean towards `OwnershipLostError` inheriting from `AzureError`.

Reply:
Well, even that is breaking:
```python
try:
   # pass
except AzureError:
   # do something
except Exception:
   # do else
```

Since after adding `AzureError` to any exceptions, you change the path.

So I'm torn to the right approach.... @johanste ?

Reply:
Isn't this exception for the inner workings of the EPH? In other words, where would a user try to catch (`except`) this type?

Reply:
User would catch this exception in the `on_error` callback if provided. This exception would be throwed during the [load-balancing stage](https://github.com/Azure/azure-sdk-for-python/blob/adfce6bcd0e66b9d02f2354c89b6fb94c7feb73f/sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/event_processor.py#L266-L286).

Reply:
Catch or get it passed in?